src/main/java/net/openhft/chronicle/core/threads/ThreadLock.java

```diff
@@ -0,0 +1,134 @@
+package net.openhft.chronicle.core.threads;
+
+import net.openhft.chronicle.core.Jvm;
+import net.openhft.chronicle.core.OS;
+import net.openhft.chronicle.core.values.LongValue;
+
+import java.io.File;
+
+public class ThreadLock {
+    static final Metrics METRICS = new Metrics();
+    private static final long TOP_BITS = ~0L << 32;
+    private static final int UNLOCKED = 0;
+    private final LongValue twoThreadId;
+    private final long timeoutMs;
+    int busyLoopCount = 20_000;
+    int busyLockSlowerCount = 100;
+
+    public ThreadLock(LongValue twoThreadId, long timeoutMs) {
+        this.twoThreadId = twoThreadId;
+        this.timeoutMs = timeoutMs;
+    }
+
+    public boolean tryLock(int threadId) {
+        if (threadId == 0)
+            throw new IllegalArgumentException();
+        long twoThreadId0 = twoThreadId.getVolatileValue();
+        int threadId0 = (int) twoThreadId0;
+        if (threadId0 != UNLOCKED) {
+            if (threadId0 == threadId)
+                throw new IllegalStateException("trying to lock twice");
+            return false;
+        }
+        return twoThreadId.compareAndSwapValue(twoThreadId0, (twoThreadId0 & TOP_BITS) | threadId);
+    }
+
+    public void lock(int threadId) throws InterruptedRuntimeException {
+        if (tryLock(threadId))
+            return;
+        busyLock(threadId);
+    }
+
+    private void busyLock(int threadId) {
+        // busy wait
+        for (int i = 0; i < busyLoopCount; i++) {
+            if (tryLock(threadId))
+                return;
+            Jvm.nanoPause();
+        }
+        if (busyLockSlower(threadId))
+            return;
+        forceUnlockAndRetry(threadId);
+    }
+
+    @SuppressWarnings("java:S3776")
+    private boolean busyLockSlower(int threadId) throws InterruptedRuntimeException {
+        long endMS = System.currentTimeMillis() + timeoutMs;
+        do {
+            for (int i = 0; i < busyLockSlowerCount; i++) {
+                if (tryLock(threadId))
+                    return true;
+                Thread.yield();
+            }
+            if (METRICS.supportsProc) {
+                int lockedThread = (int) twoThreadId.getVolatileValue();
+                if (lockedThread == 0)
+                    continue;
+                if (!isThreadRunning(lockedThread)) {
+                    Jvm.warn().on(getClass(), "ThreadId " + lockedThread + " died while holding a lock");
+                    return false;
+                }
+            }
+            if (Thread.currentThread().isInterrupted())
+                throw new InterruptedRuntimeException();
+
+        } while (System.currentTimeMillis() < endMS);
+        return false;
+    }
+
+    private boolean isThreadRunning(int lockedThread) {
+        return new File("/proc/" + lockedThread).exists();
```

Comment:
Is there a way of doing this on platforms other than Linux?

Reply:
It would be nice to know. Possibly a posix or system call.